components/lib/dataprotect/src/main/java/mozilla/components/lib/dataprotect/SecurePrefsReliabilityExperiment.kt

```diff
@@ -0,0 +1,151 @@
+ * A special "experiment" Fact will be emitted in case of an unexpected failure.
+ *
+ * Consumers of this experiment are expected to inspect emitted Facts (e.g. record them into telemetry).
+ */
+class SecurePrefsReliabilityExperiment(private val context: Context) {
+    companion object {
+        const val PREFS_NAME = "KsReliabilityExp"
+        const val PREF_DID_STORE_VALUE = "valueStored"
+        const val SECURE_PREFS_NAME = "KsReliabilityExpSecure"
+        const val PREF_KEY = "expKey"
+        const val PREF_VALUE = "some long, mildly interesting string we'd like to store"
+
+        object Actions {
+            const val EXPERIMENT = "experiment"
+            const val GET = "get"
+            const val WRITE = "write"
+            const val RESET = "reset"
+        }
+
+        @Suppress("MagicNumber")
+        enum class Values(val v: Int) {
+            SUCCESS_MISSING(1),
+            SUCCESS_PRESENT(2),
+            FAIL(3),
+            LOST(4),
+            CORRUPTED(5),
+            PRESENT_UNEXPECTED(6),
+            SUCCESS_WRITE(7),
+            SUCCESS_RESET(8)
+        }
+    }
+
+    private val securePrefs by lazy { SecureAbove22Preferences(context, SECURE_PREFS_NAME) }
+
+    private fun prefs(): SharedPreferences {
+        return context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
+    }
+
+    /**
+     * Runs an experiment. This will emit one or more [Fact]s describing results.
+     */
+    @Suppress("TooGenericExceptionCaught", "ComplexMethod")
+    operator fun invoke() {
+        try {
+            val storedVal = try {
+                securePrefs.getString(PREF_KEY)
+            } catch (e: Exception) {
+                emitFact(Actions.GET, Values.FAIL, mapOf("javaClass" to e.nameForTelemetry()))
+
+                // should this return? or proceed to the write part..?
+                return
+            }
+
+            val valueAlreadyPersisted = prefs().getBoolean(PREF_DID_STORE_VALUE, false)
+
+            val getResult = when {
+                // we didn't store the value yet, and didn't get anything back either
+                (!valueAlreadyPersisted && storedVal == null) -> {
+                    Values.SUCCESS_MISSING
+                }
+                // we got back the value we stored
+                (valueAlreadyPersisted && storedVal == PREF_VALUE) -> {
+                    Values.SUCCESS_PRESENT
+                }
+                // value was lost
+                (valueAlreadyPersisted && storedVal == null) -> {
+                    Values.LOST
+                }
+                // we got some value back, but not what we stored
+                (valueAlreadyPersisted && storedVal != PREF_VALUE) -> {
+                    Values.CORRUPTED
+                }
+                // we didn't store the value yet, but got something back either way
+                else -> {
+                    Values.PRESENT_UNEXPECTED
```

Comment:
Of course this will actually happen in production :D

Reply:
Kinda curious to see, but yeah, fairly sure it will even though HOW